### PR TITLE
test coverage

### LINKED ISSUE #93: [test coverage] Reach 100% test coverage

### PR DESCRIPTION
**Pull request by gr2m**
closes #93

- [ ] update https://github.com/eHealthAfrica/kazana/issues/131 when merged

### CODE REVIEW

**Comment by gr2m** [comment]
this is ready \o/ @jo @quesebifurcan

File: tests/integration/couchdb-auth-test.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+    server.inject({
+      method: 'GET',
+      url: '/auth-test',
+      headers: {
+        authorization: 'Bearer 123'
+      }
+    }, function (response) {
+      t.is(couch.pendingMocks[0], undefined, 'All CouchDB mocks satisfied')
+      t.is(response.statusCode, 401, 'respons with 401 status')
```
Comment (emig) [inline tests/integration/couchdb-auth-test.js]: s/respons/responds/   otherwise is correct in catalan https://en.wiktionary.org/wiki/respons

Comment (gr2m) [inline tests/integration/couchdb-auth-test.js]: it’s a [response](https://en.wiktionary.org/wiki/response), a subject, not a verb

Comment (gr2m) [inline tests/integration/couchdb-auth-test.js]: args oops nevermind me :)

Comment (gr2m) [inline tests/integration/couchdb-auth-test.js]: fixed :)

File: index.js [hunk 2]
```diff
@@ -179,19 +182,19 @@ function startKazanaServer (main, options, callback) {
...
       var stop = server.stop.bind(server)
       server.stop = function (callback) {
         pouchDbServer.stop(function () {
           stop(callback)
         })
       }
+    }
+
+    server.register(hapiPlugins, function (error) {
+      if (error) {
```
Comment (jo) [inline index.js]: Should this branch also be ignored for coverage?
```
/* istanbul ignore if */
```

**Comment by jo** [comment]
Tests pass with bravour:
```
-----------------------------|----------|----------|----------|----------|----------------|
File                         |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
-----------------------------|----------|----------|----------|----------|----------------|
 __root__/                   |      100 |      100 |      100 |      100 |                |
  index.js                   |      100 |      100 |      100 |      100 |                |
 lib/                        |      100 |      100 |      100 |      100 |                |
  compose-plugin.js          |      100 |      100 |      100 |      100 |                |
  gather-plugins.js          |      100 |      100 |      100 |      100 |                |
  log.js                     |      100 |      100 |      100 |      100 |                |
 lib/auth/                   |      100 |      100 |      100 |      100 |                |
  couchdb.js                 |      100 |      100 |      100 |      100 |                |
 lib/decorators/             |      100 |      100 |      100 |      100 |                |
  reply-pouchdb-error.js     |      100 |      100 |      100 |      100 |                |
 lib/methods/                |      100 |      100 |      100 |      100 |                |
  get-couch-url.js           |      100 |      100 |      100 |      100 |                |
  get-store.js               |      100 |      100 |      100 |      100 |                |
  get-user.js                |      100 |      100 |      100 |      100 |                |
 lib/plugins/                |      100 |      100 |      100 |      100 |                |
  logger.js                  |      100 |      100 |      100 |      100 |                |
 lib/utils/                  |      100 |      100 |      100 |      100 |                |
  token-from-request.js      |      100 |      100 |      100 |      100 |                |
  token-to-cookie-headers.js |      100 |      100 |      100 |      100 |                |
-----------------------------|----------|----------|----------|----------|----------------|
All files                    |      100 |      100 |      100 |      100 |                |
-----------------------------|----------|----------|----------|----------|----------------|
```

:clap: :clap: :clap:

File: lib/compose-plugin.js [hunk 3]
```diff
@@ -159,7 +175,12 @@ function createOrUpdateIndices (server, indicesMaps, callback) {
     return indices.map(store.updateOrAdd)
   })
 
-  server.app.utils.Promise.all(promises)
+  // flatten
+  promises = promises.reduce(function (a, b) {
```
Comment (jo) [inline lib/compose-plugin.js]: Since we already have lodash in place should'nt we use its `flatten()` function?

File: lib/methods/get-couch-url.js [hunk 4]
```diff
@@ -7,11 +7,11 @@ module.exports = getCouchUrl
  * @returns  {Promise}
  **/
 function getCouchUrl () {
-  var kazana = this
+  var server = this
```
Comment (jo) [inline lib/methods/get-couch-url.js]: :+1:

File: lib/methods/get-store.js [hunk 5]
```diff
@@ -44,23 +44,23 @@ function toDbName (kazana, name) {
...
+  }
 
   if (options.auth === 'admin') {
     return _.defaults({
-      auth: kazana.settings.app.admin
+      auth: server.settings.app.admin
     }, ajaxOptions)
   }
 
-  Hoek.assert(options.auth, 'invalid options.auth for getStore')
```
Comment (jo) [inline lib/methods/get-store.js]: Why was this line removed? Is this because we have defaults dor `settings.app.db.options`?

File: lib/methods/get-user.js [hunk 6]
```diff
@@ -2,10 +2,11 @@ module.exports = getUser
 
 var _ = require('lodash')
 var Boom = require('boom')
-var PouchDB = require('pouchdb')
-var Promise = require('pouchdb/extras/promise')
 var Hoek = require('hoek')
-PouchDB.plugin(require('pouchdb-authentication'))
+
+var internals = module.exports.internals = {}
```
Comment (jo) [inline lib/methods/get-user.js]: Don't know if I like this _export for mocking_ thing...

File: package.json [hunk 7]
```diff
@@ -30,7 +31,7 @@
     "npm": ">=2.14.0"
   },
   "dependencies": {
-    "async": "^1.5.0",
+    "async": "git://github.com/gr2m/async",
```
Comment (jo) [inline package.json]: I think your pull landed in v1.5.1

File: package.json [hunk 8]
```diff
@@ -5,6 +5,7 @@
   "version": "0.0.0-semantic-released",
   "scripts": {
     "coverage": "nyc report --reporter=text-lcov | coveralls",
+    "coverage:local": "nyc report --reporter=lcov",
     "pretest": "standard",
     "test": "nyc tap --no-cov tests/**/*-test.js",
```
Comment (jo) [inline package.json]: From what I understand nyc is included in tap and we can print coverage and [upload to coveralls.io](https://github.com/tapjs/node-tap/blob/master/CHANGELOG.md#31) with tap itself?

**Comment by jo** [comment]
Rebased

**Comment by jo** [comment]
LGTM very well done!

File: tests/integration/couchdb-auth-test.js [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    kazana: {
+      log: {
+        level: 'error'
+      },
+      backend: {
+        name: 'couchdb',
+        location: 'http://localhost:5984'
+      },
+      admin: {
+        usrname: 'admin',
```
Comment (jo) [inline tests/integration/couchdb-auth-test.js]:
```diff
-usrname
+username
```

Comment (jo) [inline tests/integration/couchdb-auth-test.js]: That it worked is a sign that the whole admin section can be removed, right?

File: package.json (hunk 8 above)
Comment (jo) [inline package.json]: I simplified the coverage system in https://github.com/eHealthAfrica/kazana-bootstrap/commit/7cced8565179f146acd5f7009cdbb4a59f23788e

However, we now need to add the `COVERALLS_REPO_TOKEN` environment variable but I opened an issue for that on tap: https://github.com/tapjs/node-tap/issues/210

File: lib/compose-plugin.js (hunk 3 above)
Comment (gr2m) [inline lib/compose-plugin.js]: yeah why not :)

File: lib/methods/get-store.js (hunk 5 above)
Comment (gr2m) [inline lib/methods/get-store.js]: Can’t recall why, we can add it back in and see if tests fail?

File: lib/methods/get-user.js (hunk 6 above)
Comment (gr2m) [inline lib/methods/get-user.js]: it’s just internal, the package does not expose it. And it works very vell for unit tests

File: package.json (hunk 7 above)
Comment (gr2m) [inline package.json]: yes

File: package.json (hunk 8 above)
Comment (gr2m) [inline package.json]: I think the `COVERALLS_REPO_TOKEN` environment variable is not needed on travis. 

I’ve read that tap comes with tap the other day, but didn’t test it yet. If it works, cool :)

File: tests/integration/couchdb-auth-test.js (hunk 9 above)
Comment (gr2m) [inline tests/integration/couchdb-auth-test.js]: yep :)

File: package.json (hunk 8 above)
Comment (jo) [inline package.json]: It is needed, I have tried it here: https://travis-ci.org/eHealthAfrica/kazana-bootstrap/builds/100555951
If you remove the token it does not publish to coverall.io

File: lib/compose-plugin.js (hunk 3 above)
Comment (jo) [inline lib/compose-plugin.js]: done

File: lib/methods/get-store.js (hunk 5 above)
Comment (jo) [inline lib/methods/get-store.js]: done. Tests don't fail

File: package.json (hunk 7 above)
Comment (jo) [inline package.json]: done

File: package.json (hunk 8 above)
Comment (jo) [inline package.json]: Lets do this in a separate pull

File: tests/integration/couchdb-auth-test.js (hunk 9 above)
Comment (jo) [inline tests/integration/couchdb-auth-test.js]: removed that admin section, tests still pass.
